src/bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,378 @@
+      parameters (the handler's job);
+    - Interpreting those DataSource and Asset parameters to do I/O (the adapter's job).
+"""
+
+# The below is confusing (to Dan).
+# We have a use for StreamDatum(s) in, data out.
+# No dependency on Tiled, its database, HTTP, etc.
+# We just want fast incremental access to streaming data.
+
+# In common:
+# - knowledge of format details (meaning of parameters like swmr)
+# - instance state to track things like num_total_rows.
+
+# Different:
+# - The "at rest" use case has a conceptual dependency on
+#   DataSource Asset.
+# - The streaming data use case has an import dependency o
+#   whatever the I/O library is (e.g. h5py).
+
+
+class Thing:
+    def __init__(self, data_uri, swmr):  # TODO Add reshape, skips.
+        ...
+
+    def read_stream_datum(self, *stream_datum_docs):
+        array = ...
+        return array
+
+
+class StreamHandlerBase:
+    mimetype: str = ""
+    _num_rows: int = 0
+
+    def __init__(self):
+        pass
+
+    @abstractmethod
+    def consume_stream_datum(self, doc):
+        # This will be called for every new StreamDatum received.
+        # Actions:
+        #   - Consume new StreamDatum
+        #   - Update shape and chunks
+        pass
+
+
+class HDF5StreamHandler(StreamHandlerBase):
+    mimetype = "application/x-hdf5"
+
+    def __init__(
+        self, data_uri, path, data_shape, data_type, chunk_size: Optional[int] = None, swmr=True, **kwargs
+    ):
+        self.assets = [Asset(data_uri=data_uri, is_directory=False, parameter="data_uri")]
+        self.path = path.strip("/").split("/")
+        self.data_shape = data_shape  # Intrinsic data shape (single frame as written in HDF5 file)
+        self.data_type = np.dtype(data_type)
+        self.chunk_size = chunk_size
+        self.adapter_parameters = {"path": self.path}
+
+    @property
+    def shape(self):
+        """Shape including the 0-th dimension -- number of rows"""
+        # TODO Add reshape, skips.
+        return [self._num_rows, *self.data_shape]
+
+    @property
+    def chunks(self):
+        """Chunking specification based on the Stream Resource parameter `chunk_size`:
+        None or 0 -- single chunk for all existing and new elements
+        int -- fixed-sized chunks with at most `chunk_size` elements, last chunk can be smaller
+        """
+        if not self.chunk_size:
+            dim0_chunk = [self._num_rows]
+        else:
+            dim0_chunk = [self.chunk_size] * int(self._num_rows / self.chunk_size)
+            if self._num_rows % self.chunk_size:
+                dim0_chunk.append(self._num_rows % self.chunk_size)
+
+        return [dim0_chunk] + [[d] for d in self.data_shape]
+
+    def consume_stream_datum(self, doc):
+        """Append StreamDatum to an existing Handler (StreamResource)"""
+        # TODO Add reshape, skips.
+        self._num_rows += doc["indices"]["stop"] - doc["indices"]["start"]  # Number of rows added by StreamDatum
```

Comment:
AFAIK _num_rows should be doc["seq_nums"] stop - start, the data pointed to within the h5 file is the indices

Reply:
That's right, but `_num_rows` here corresponds to the total number of rows in the Data Source (file), including skips. I have added a comment to clarify this.